app/upgrades_test.go

```diff
@@ -0,0 +1,56 @@
+package app_test
+
+import (
+	"testing"
+	"time"
+
+	"cosmossdk.io/log"
+	upgradetypes "cosmossdk.io/x/upgrade/types"
+	"github.com/celestiaorg/celestia-app/v6/app"
+	"github.com/celestiaorg/celestia-app/v6/test/util"
+	"github.com/celestiaorg/celestia-app/v6/test/util/testfactory"
+	tmdb "github.com/cosmos/cosmos-db"
+	"github.com/cosmos/cosmos-sdk/baseapp"
+	"github.com/stretchr/testify/require"
+)
+
+func TestUpgrades(t *testing.T) {
+	t.Run("app.New() should register a v6 upgrade handler", func(t *testing.T) {
+		logger := log.NewNopLogger()
+		db := tmdb.NewMemDB()
+		traceStore := &NoopWriter{}
+		timeoutCommit := time.Second
+		appOptions := NoopAppOptions{}
+
+		testApp := app.New(logger, db, traceStore, timeoutCommit, appOptions, baseapp.SetChainID(testfactory.ChainID))
+
+		require.False(t, testApp.UpgradeKeeper.HasHandler("v5"))
+		require.True(t, testApp.UpgradeKeeper.HasHandler("v6"))
+	})
+}
+
+func TestApplyUpgrade(t *testing.T) {
+	t.Run("ICA host params should have an explicit allowlist of messages before and after ApplyUpgrade", func(t *testing.T) {
+		testApp, _, _ := util.NewTestAppWithGenesisSet(app.DefaultConsensusParams())
+		require.True(t, testApp.UpgradeKeeper.HasHandler("v6"))
+		plan := upgradetypes.Plan{
+			Name:   "v6",
+			Time:   time.Now(),
+			Height: 1,
+			Info:   "info",
+		}
+
+		ctx := testApp.NewContext(false)
+		params := testApp.ICAHostKeeper.GetParams(ctx)
+		require.True(t, params.HostEnabled)
+		require.Equal(t, params.AllowMessages, app.IcaAllowMessages())
+
+		err := testApp.UpgradeKeeper.ApplyUpgrade(ctx, plan)
+		require.NoError(t, err)
+
+		ctx = testApp.NewContext(false)
+		params = testApp.ICAHostKeeper.GetParams(ctx)
+		require.True(t, params.HostEnabled)
+		require.Equal(t, params.AllowMessages, app.IcaAllowMessages())
```

Comment:
blocking question

will this test run the upgrade or do we need to change the default consensus params to v5 first?

https://github.com/rootulp/celestia-app/blob/cfd9c4112c5b1b94d1dcbbe8607c37d452eed44c/app/default_overrides.go#L238

Reply:
This test runs the upgrade because it executes the fix (`setICAHostParams`)

Reply:
if it requires running the multiplexer to run v5 then upgrade we can just test in the upgrade e2e test

Reply:
The params are the same before and after.

This doesn't repro what's happening on Mainnet b/c we expect the params to be unset on Mainnet (and/or the module to not be registered) before the upgrade.

> if it requires running the multiplexer to run v5 then upgrade we can just test in the upgrade e2e test imo

Agreed. Tracking in https://github.com/celestiaorg/celestia-app/issues/5452

> or maybe we could remove the params, apply the upgrade, then check to see if they're set if we want a unit test?

I can remove or modify this unit test considering it's causing more confusing than helping. It was helpful at development time because the IBC param migrator does not work as expected so we needed to write our own.